sound/soc/sof/pm.c

```diff
@@ -500,3 +500,58 @@ void snd_sof_d0i3_entry_work(struct work_struct *work)
 	}
 }
 EXPORT_SYMBOL(snd_sof_d0i3_entry_work);
+
+/**
+ * sof_set_dsp_d0i0 - bring ADSP back to d0i0
+ * @force: force to exit from d0i3
+ * @reload: reload unused pipelines/widgets into FW
+ */
+void snd_sof_set_dsp_d0i0(struct snd_sof_dev *sdev, bool force, bool reload)
+{
+	int ret;
+
+	if (sdev->d0_state == SOF_DSP_D0I0) {
+		dev_vdbg(sdev->dev, "Already in D0i0.\n");
+		return;
+	}
+
+	if (sof_d0ix_target_state(sdev) == SOF_DSP_D0I3 &&
+	    !force) {
+		dev_dbg(sdev->dev, "Target is D0i3, not forcing it to D0i0\n");
+		return;
+	}
+
+	dev_dbg(sdev->dev, "exiting from D0I3.\n");
+	/* cancel any pending d0i3 entry work */
+	cancel_delayed_work_sync(&sdev->d0i3_entry_work);
+
+	/* do platform specific d0i3 exit */
+	snd_sof_dsp_d0ix_update(sdev, false);
+
+	/* sending d0i3_exit IPC */
+	ret = sof_send_pm_ipc(sdev, SOF_IPC_PM_D0I3_EXIT);
+	if (ret < 0)
+		dev_err(sdev->dev,
+			"error: d0i3_exit ipc error %d\n", ret);
+
+	/* update dsp state */
+	sdev->d0_state = SOF_DSP_D0I0;
+}
+EXPORT_SYMBOL(snd_sof_set_dsp_d0i0);
+
+void snd_sof_set_dsp_d0i3(struct snd_sof_dev *sdev)
+{
+	if (sdev->d0_state == SOF_DSP_D0I3) {
+		dev_dbg(sdev->dev, "Already in D0i3.\n");
+		return;
+	}
+
+	/* Schedule D0i3 entry work if needed */
+	if (sof_d0ix_target_state(sdev) == SOF_DSP_D0I3) {
+		dev_dbg(sdev->dev, "will schedule D0i3 work after %d ms\n",
+			SND_SOF_D0I3_DELAY);
+		schedule_delayed_work(&sdev->d0i3_entry_work,
+				      msecs_to_jiffies(SND_SOF_D0I3_DELAY));
```

Comment:
if the helper is called, you'd need to cancel previous calls to 'schedule' so that the timer is reset to the last call to this helper (same as mark_last_busy for pm_runtime).

Reply:
Wondering: this looks very similar to automatic delayed runtime-PM, can we maybe somehow reuse that API here? If we don't have a "free" device, to which we could meaningfully attach this RT-PM, maybe even add a new device for this.

Reply:
> if the helper is called, you'd need to cancel previous calls to 'schedule' so that the timer is reset to the last call to this helper (same as mark_last_busy for pm_runtime).

It looks not easy to handle this decently, as @lyakh suggested, let me consider reusing runtime-PM APIs.